corehq/motech/dhis2/tests/test_entities_helpers.py

```diff
@@ -437,6 +440,122 @@ def test_existing_relationship_not_created_again(self):
+        cls.factory = CaseFactory(domain=DOMAIN)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.domain_obj.delete()
+        super().tearDownClass()
+
+    def setUp(self):
+        build_patch = patch('corehq.motech.dhis2.entities_helpers._build_tracked_entity')
+        self.build_func = build_patch.start()
+        self.addCleanup(build_patch.stop)
+        self.build_func.return_value = [{'trackedEntityInstance': 'foobar'}, None]
+
+        post_patch = patch('corehq.motech.requests.Requests.post')
+        self.post_func = post_patch.start()
+        self.addCleanup(post_patch.stop)
+        self.post_func.return_value = Mock(
+            status_code=200,
+            json=lambda: {
+                'response': {
+                    'importSummaries': ['foobar'],
+                    'enrollments': {
+                        'importSummaries': [
+                            {
+                                'events': {
+                                    'importSummaries': [
+                                        {
+                                            'status': 'ERROR',
+                                            'description': 'foobar error'
+                                        }
+                                    ]
+                                }
+                            }
+                        ]
+                    }
+                }
+            },
+        )
+
+    def test_register_tracked_entity_instance_errors(self):
+        requests = Requests(
+            DOMAIN,
+            'https://dhis2.example.com/',
+            auth_manager=AuthManager(),
+        )
+        case_trigger_info = CaseTriggerInfo(domain=DOMAIN, case_id=None)
+        case_config = {
+            'te_type_id': 'foobar',
+            'org_unit_id': {
+                'doc_type': 'CaseOwnerAncestorLocationField',
+                'location_field': 'dhis_id'
+            },
+            'tei_id': 'foobar'
+        }
+
+        with self.assertRaises(Dhis2Exception):
+            register_tracked_entity_instance(requests, case_trigger_info, case_config)
+
+
+class TestUpdateTrackedEntityInstance(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.domain_obj = create_domain(DOMAIN)
+        cls.factory = CaseFactory(domain=DOMAIN)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.domain_obj.delete()
+        super().tearDownClass()
+
+    def setUp(self):
+        build_patch = patch('corehq.motech.dhis2.entities_helpers._build_tracked_entity')
+        self.build_func = build_patch.start()
+        self.addCleanup(build_patch.stop)
+        self.build_func.return_value = [{"trackedEntityInstance": "foobar"}, None]
+
+        put_patch = patch('corehq.motech.requests.Requests.put')
+        self.put_func = put_patch.start()
+        self.addCleanup(put_patch.stop)
+        self.put_func.return_value = Mock(
+            status_code=200,
```

Comment:
I know we're using integers as status codes in _many_ places in the code, but this [isn't recommended](https://www.django-rest-framework.org/api-guide/status-codes/#:~:text=Using%20bare%20status%20codes%20in%20your%20responses%20isn%27t%20recommended). We should rather be using (and start using) constants. See [this](https://www.django-rest-framework.org/api-guide/status-codes/) page talking about exactly this. I feel like making a tech dept ticket and tackling all status codes in the codebase..but maybe that's a bit ambitious. We got to start somewhere right? :sweat_smile:

Reply:
Ah I saw those! But where I was working, all the other functions made use of integers so I figured to keep that consistency. But that's a really good idea, it would make identifying what the status code refers to a lot easier! 

I think a good place to start would be to update the status codes in the file or part of the code you're working on, and then they'll slowly get replaced like that. :thinking:

Reply:
We should then keep the refactors in a separate commit also, but yes, I agree.

Reply:
Maybe I'm in a quibbly mood.

In general, I would definitely advocate using constants. They make it easy to find usage, and much (much) easier to refactor or modify.

_However_ the reasoning that the DRF documentation gives for using constants for HTTP status codes is "... to make your code more obvious and readable." :thinking: :thinking: :thinking: No. In my very opinionated opinion, nothing is more obvious and readable than an integer value when you're expecting an integer value. And if the IETF / some new Emperor of the Internet decides to change HTTP status codes to different values, then we have bigger problems than grepping for "200".

I think in the case of HTTP status codes, integer values are 200 by me. :smirk: